client/alloc_runner_health_watcher.go

```diff
@@ -0,0 +1,203 @@
+		return
+	}
+
+	tg := alloc.Job.LookupTaskGroup(alloc.TaskGroup)
+	if tg == nil {
+		r.logger.Printf("[ERR] client.alloc_watcher: failed to lookup allocation's task group. Exiting watcher")
+		return
+	}
+
+	// Checks marks whether we should be watching for Consul health checks
+	desiredChecks := 0
+	var checkTicker *time.Ticker
+	var checkCh <-chan time.Time
+
+	u := tg.Update
+	switch {
+	case u == nil:
+		r.logger.Printf("[TRACE] client.alloc_watcher: no update block for alloc %q. exiting", alloc.ID)
+		return
+	case u.HealthCheck == structs.UpdateStrategyHealthCheck_Manual:
+		r.logger.Printf("[TRACE] client.alloc_watcher: update block has manual checks for alloc %q. exiting", alloc.ID)
+		return
+	case u.HealthCheck == structs.UpdateStrategyHealthCheck_Checks:
+		for _, task := range tg.Tasks {
+			for _, s := range task.Services {
+				desiredChecks += len(s.Checks)
+			}
+		}
+
+		checkTicker = time.NewTicker(consulCheckLookupInterval)
+		checkCh = checkTicker.C
+	}
+
+	// Get a listener so we know when an allocation is updated.
+	l := r.allocBroadcast.Listen()
+
+	// Create a deadline timer for the health
+	deadline := time.NewTimer(u.HealthyDeadline)
+
+	// Create a healthy timer
+	latestTaskHealthy := time.Unix(0, 0)
+	latestChecksHealthy := time.Unix(0, 0)
+	healthyTimer := time.NewTimer(0)
+	if !healthyTimer.Stop() {
+		<-healthyTimer.C
+	}
+
+	// Cleanup function
+	defer func() {
+		if !deadline.Stop() {
+			<-deadline.C
+		}
+		if !healthyTimer.Stop() {
+			<-healthyTimer.C
+		}
+		if checkTicker != nil {
+			checkTicker.Stop()
+		}
+		l.Close()
+	}()
+
+	setHealth := func(h bool) {
+		r.allocLock.Lock()
+		r.allocHealth = helper.BoolToPtr(h)
+		r.allocLock.Unlock()
+		r.syncStatus()
+	}
+
+	var checks []*api.AgentCheck
+	first := true
+OUTER:
+	for {
+		if !first {
+			select {
+			case <-ctx.Done():
+				return
+			case newAlloc, ok := <-l.Ch:
+				if !ok {
+					return
+				}
+
+				alloc = newAlloc
+				r.logger.Printf("[TRACE] client.alloc_watcher: new alloc version for %q", alloc.ID)
+			case <-checkCh:
+				newChecks, err := r.consulClient.Checks(alloc)
+				if err != nil {
+					r.logger.Printf("[TRACE] client.alloc_watcher: failed to lookup consul checks for allocation %q: %v", alloc.ID, err)
+				}
+
+				checks = newChecks
+			case <-deadline.C:
+				// We have exceeded our deadline without being healthy.
+				r.logger.Printf("[TRACE] client.alloc_watcher: alloc %q hit healthy deadline", alloc.ID)
+				setHealth(false)
+				return
+			case <-healthyTimer.C:
+				r.logger.Printf("[TRACE] client.alloc_watcher: alloc %q is healthy", alloc.ID)
+				setHealth(true)
+				return
+			}
+		}
+		first = false
+
+		// If the alloc is being stopped by the server just exit
+		switch alloc.DesiredStatus {
+		case structs.AllocDesiredStatusStop, structs.AllocDesiredStatusEvict:
+			r.logger.Printf("[TRACE] client.alloc_watcher: desired status terminal for alloc %q", alloc.ID)
+			return
```

Comment:
Nope since the nil case means it hasn't been determined, which is valid. It shut down before becoming healthy/unhealthy.